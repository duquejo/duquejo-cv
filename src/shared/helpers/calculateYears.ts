import { computed } from 'vue';
import { START_ADVENTURE_YEAR } from '@/shared/constants';

export const calculateYears = computed(() => (new Date().getFullYear() - START_ADVENTURE_YEAR));