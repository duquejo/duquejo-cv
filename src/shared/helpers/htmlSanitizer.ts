export const htmlSanitizer = (html: string) => {
  return html.replace(/<\/?(?!\/?(b|strong|em|i|u|s|br)\b)[^>]*>/gi, '');
};