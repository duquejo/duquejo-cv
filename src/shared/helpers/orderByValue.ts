
export const orderObjectListByKey = <T>( list: T[], key: keyof T ): T[] | null => {
  if( ! list || list.length === 0 ) return null;
  return list.sort((a: T, b: T) => parseInt(String(b[key])) - parseInt(String(a[key])) );
}