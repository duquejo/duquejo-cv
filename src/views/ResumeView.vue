<template>
  <WrapperLayout>
    <section class="row items-center mx-0 mt-0 px-0">
      <div class="w-full order-2 lg:basis-1/2 lg:order-1 px-8 pt-8 lg:m-auto">
        <span class="text-gray-500 text-lg">{{ t('general.role') }}</span>
        <h1 class="mt-2">{{ t('general.resume.greetings') }} <strong>José Duque.</strong></h1>
        <p v-html="htmlSanitizer(t('general.resume.excerpt'))" />
        <div class="button-container">
          <v-icon v-for="stack in TECH_STACK" class="hover:animate-wiggle" :name="stack.icon" scale="2.5"
            :title="stack.title" />
        </div>
      </div>
      <picture class="w-full order-1 lg:basis-1/2 lg:order-2 overflow-hidden">
        <img class="shadow transition-all duration-500 scale-110 hover:rotate-3"
          :class="isPortraitLoaded ? 'opacity-100' : 'opacity-0'" :src="avatar" alt="José Miguel Duque" loading="lazy"
          @load="onLoaded" />
      </picture>
    </section>
    <section>
      <h2 class="text-h2" v-text="t('general.resume.description.title')" />
      <p v-html="htmlSanitizer(t('general.resume.description.content'))" />
    </section>
  </WrapperLayout>
</template>

<script setup lang="ts">
import avatar from '@/assets/avatar.webp';
import TECH_STACK from '@/shared/data/stack.json';
import useLanguageContext from '@/composables/useLanguageContext';
import WrapperLayout from '@/layouts/WrapperLayout.vue';

import { htmlSanitizer } from '@/shared/helpers/htmlSanitizer';
import { ref } from 'vue';

const isPortraitLoaded = ref<boolean>(false);
const { t } = useLanguageContext();

const onLoaded = () => {
  isPortraitLoaded.value = true;
};
</script>