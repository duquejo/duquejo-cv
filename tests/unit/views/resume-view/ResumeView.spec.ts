import { mount } from '@vue/test-utils';
import { describe, expect, it, vi } from 'vitest';

import ResumeView from '../../../../src/views/ResumeView.vue';
import createVueI18n from '../../../mocks/mock-i18n';
import * as utils from '../../../../src/shared/helpers/htmlSanitizer';

describe('+ ResumeView tests', () => {

  const i18n = createVueI18n('en');
  const htmlSanitizerSpy = vi.spyOn(utils, 'htmlSanitizer');

  const scrollIntoViewSpy = vi.fn();
  window.HTMLElement.prototype.scrollIntoView = scrollIntoViewSpy;

  it('Should match the snapshot', () => {
    const wrapper = mount(ResumeView, {
      global: {
        plugins: [i18n],
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
    expect(htmlSanitizerSpy).toHaveBeenCalled();
    expect(htmlSanitizerSpy).toHaveBeenCalledTimes(2);
    expect(htmlSanitizerSpy).toHaveBeenCalledWith('general.resume.excerpt2');
    expect(htmlSanitizerSpy).toHaveBeenCalledWith('general.resume.description.content');

    expect(scrollIntoViewSpy).toHaveBeenCalled();
    expect(scrollIntoViewSpy).toHaveBeenCalledWith({
      behavior: 'instant'
    });
  });

  it('should load the image after tick', async () => {
    // Arrange
    const wrapper = mount(ResumeView, {
      global: {
        plugins: [i18n],
      }
    }) as any;

    // Act & Assert
    expect(wrapper.find('img').classes('opacity-0'));
    expect(wrapper.vm.isPortraitLoaded).toBeFalsy();
    wrapper.vm.onLoaded();
    expect(wrapper.find('img').classes('opacity-100'));
    expect(wrapper.vm.isPortraitLoaded).toBeTruthy();
  });
});