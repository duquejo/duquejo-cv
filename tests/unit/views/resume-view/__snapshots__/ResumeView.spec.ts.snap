// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`+ ResumeView tests > Should match the snapshot 1`] = `
"<div class="animated-container">
  <article class="mt-5 lg:mt-0">
    <section class="row items-center mx-0 mt-0 px-0">
      <div class="w-full order-2 lg:basis-1/2 lg:order-1 px-8 pt-8 lg:m-auto"><span class="text-gray-500 text-lg">general.role</span>
        <h1 class="mt-2">general.resume.greetings <strong>José Duque.</strong></h1>
        <p>general.resume.excerpt1</p>
        <p>general.resume.excerpt2</p>
        <div class="button-container">
          <v-icon class="hover:animate-wiggle" name="vi-file-type-typescript-official" scale="2.5" title="Typescript"></v-icon>
          <v-icon class="hover:animate-wiggle" name="vi-file-type-node" scale="2.5" title="Node JS"></v-icon>
          <v-icon class="hover:animate-wiggle" name="vi-file-type-tailwind" scale="2.5" title="Tailwind"></v-icon>
          <v-icon class="hover:animate-wiggle" name="vi-file-type-vue" scale="2.5" title="Vue JS"></v-icon>
          <v-icon class="hover:animate-wiggle" name="vi-file-type-reactjs" scale="2.5" title="React JS"></v-icon>
        </div>
      </div>
      <picture class="w-full order-1 lg:basis-1/2 lg:order-2 overflow-hidden"><img class="shadow transition-all duration-500 scale-110 hover:rotate-3 aspect-[200/170] opacity-0" alt="José Miguel Duque" sizes="(max-width: 1400px) 100vw, 1400px" srcset="/static/img/avatar_w_200.webp 200w, /static/img/avatar_w_558.webp 558w, /static/img/avatar_w_825.webp 825w, /static/img/avatar_w_1089.webp 1089w, /static/img/avatar_w_1255.webp 1255w, /static/img/avatar_w_1400.webp 1400w" src="/static/img/avatar_w_1400.webp"></picture>
    </section>
    <section>
      <h2 class="text-h2">general.resume.description.title</h2>
      <p>general.resume.description.content</p>
    </section>
  </article>
  <!---->
</div>"
`;
